cookbooks/aws-parallelcluster-scheduler-plugin/files/default/clusterstatusmgtd/clusterstatusmgtd.py

```diff
@@ -0,0 +1,450 @@
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import json
+import logging
+import os
+import shlex
+import subprocess
+import time
+from configparser import ConfigParser
+from datetime import datetime, timezone
+from enum import Enum
+from logging.config import fileConfig
+
+import boto3
+from boto3.dynamodb.conditions import Attr
+from botocore.config import Config
+
+CONFIG_FILE_DIR = "/etc/parallelcluster/scheduler_plugin"
+LOOP_TIME = 60
+log = logging.getLogger(__name__)
+
+
+# Utils
+def _seconds(sec):
+    """Convert seconds to milliseconds."""
+    return sec * 1000
+
+
+def _sleep_remaining_loop_time(total_loop_time, loop_start_time=None):
+    end_time = datetime.now(tz=timezone.utc)
+    if not loop_start_time:
+        loop_start_time = end_time
+    # Always convert the received loop_start_time to utc timezone. This is so that we never rely on the system local
+    # time and risk to compare native datatime instances with localized ones
+    loop_start_time = loop_start_time.astimezone(tz=timezone.utc)
+    time_delta = (end_time - loop_start_time).total_seconds()
+    if 0 <= time_delta < total_loop_time:
+        time.sleep(total_loop_time - time_delta)
+
+
+def log_exception(
+    logger,
+    action_desc,
+    log_level=logging.ERROR,
+    catch_exception=Exception,
+    raise_on_error=True,
+    exception_to_raise=None,
+):
+    def _log_exception(function):
+        @functools.wraps(function)
+        def wrapper(*args, **kwargs):
+            try:
+                return function(*args, **kwargs)
+            except catch_exception as e:
+                logger.log(log_level, f"Failed when {action_desc} with exception {e}")
+                if raise_on_error:
+                    if exception_to_raise:
+                        raise exception_to_raise
+                    else:
+                        raise
+
+        return wrapper
+
+    return _log_exception
+
+
+def _run_command(  # noqa: C901
+    command,
+    capture_output=True,
+    log_error=True,
+    log_file=None,
+    env=None,
+    timeout=None,
+    raise_on_error=True,
+):
+    """Execute shell command."""
+    if isinstance(command, str):
+        command = shlex.split(command)
+    log_command = command if isinstance(command, str) else " ".join(str(arg) for arg in command)
+    logging.info(f"Executing command: {log_command}")
+    try:
+        result = subprocess.run(  # nosec - trusted input
+            command,
+            capture_output=capture_output,
+            universal_newlines=True,
+            encoding="utf-8",
+            env=env,
+            timeout=timeout,
+        )
+        result.check_returncode()
+    except subprocess.CalledProcessError:
+        if log_error:
+            if log_file:
+                logging.error(f"Command {log_command} failed, check log file {log_file}")
```

Comment:
why the run_command function should be aware of log_file? I think the caller should have the logging logic pointing to the log file and this function should be called with capture_output=False.